File: typescript/Shell/play/1.loki.ts
import { style, fmt } from "../0.root.js";

export default (string: string, frames = 1) => {
  const characters = Math.floor(Math.random() * string.length);
  const styles = Object.keys(style);
  const renders = [];

  if (characters > string.length) {
    string = string.padEnd(characters, " ");
  }

  for (let i = 0; i < frames; i++) {
    const styledIndices = new Set();
    let styledString = string;
    for (let j = 0; j < characters; j++) {
      let randomIndex;
      do {
        randomIndex = Math.floor(Math.random() * string.length);
      } while (styledIndices.has(randomIndex));

      styledIndices.add(randomIndex);

      const randomStyle = styles[Math.floor(Math.random() * styles.length)];
      const styledCharacter = fmt(string[randomIndex], randomStyle);

      styledString =
        styledString.substring(0, randomIndex) +
        styledCharacter +
        styledString.substring(randomIndex + 1);
    }
    renders.push(styledString);
  }

  return renders;
};

File: gosource/shell/play/1.loki.go
package play

import (
	"math/rand"
	"reflect"
	"strings"

	"main/shell/render"
	"main/shell/root"
)

// renderLoki generates the animation frames for text styling animation
func renderLoki(s string, frames int) []string {
	// Get available style constants
	var styleFields []string
	t := reflect.TypeOf(root.Style)
	for i := 0; i < t.NumField(); i++ {
		field := t.Field(i)
		styleFields = append(styleFields, field.Name)
	}

	// Calculate random number of characters to style
	characters := rand.Intn(len(s))
	if characters > len(s) {
		s = s + strings.Repeat(" ", characters-len(s))
	}

	// Generate frames
	var renders []string
	for range frames {
		used := make(map[int]bool)
		result := []rune(s)

		for range characters {
			// Find unused random index
			var randomIndex int
			for {
				randomIndex = rand.Intn(len(s))
				if !used[randomIndex] {
					break
				}
			}
			used[randomIndex] = true

			// Get random style
			randomStyle := reflect.ValueOf(root.Style).FieldByName(styleFields[rand.Intn(len(styleFields))]).String()

			// Style the character
			styledChar := root.Format(string(result[randomIndex]), randomStyle)
			result[randomIndex] = []rune(styledChar)[0]
		}

		renders = append(renders, string(result))
	}

	return renders
}

// Loki displays an animated sequence of randomly styled text
func Loki(s string, duration, frames int) error {
	return render.Animate(renderLoki(s, frames), duration, frames)
}
